File: backend/src/main/scala/domains/judger/objects/JudgerId.scala
package domains.judger.objects

import io.circe.{Decoder, Encoder}

/** judger 租约标识；管理端 API 使用该类型展示已注册 worker。 */
final case class JudgerId(value: String)

/** JudgerId 的 JSON 编解码和格式校验入口。 */
object JudgerId:
  given Encoder[JudgerId] = Encoder.encodeString.contramap(_.value)
  given Decoder[JudgerId] = Decoder.decodeString.emap(parse)

  /** 校验 judger id，保持与 worker 协议 JudgerId 相同的长度和空值规则。 */
  def parse(raw: String): Either[String, JudgerId] =
    val normalized = raw.trim
    if normalized.isEmpty then Left("Judger id is required.")
    else if normalized.length > 120 then Left("Judger id must be at most 120 characters.")
    else Right(JudgerId(normalized))

File: frontend/src/objects/judger/JudgerId.ts
/** judger 租约标识；管理端 API 用于展示已注册 worker。 */
export type JudgerId = string & { readonly __brand: 'JudgerId' }

/** 解析 helper 的返回结果；成功携带品牌值，失败携带可展示错误，不抛异常。 */
type ParseResult<T> = { ok: true; value: T } | { ok: false; error: string }

/** 创建 judger id 品牌值；调用前必须完成空值和长度校验。 */
function createJudgerId(value: string): JudgerId {
  /** 注意：这里的 as 只在 parseJudgerId 校验通过后施加品牌类型。 */
  return value as JudgerId
}

/** 将 judger id 品牌值还原为普通字符串；用于展示、key 和 API body。 */
export function judgerIdValue(judgerId: JudgerId): string {
  return judgerId
}

/** 解析 judger id；拒绝空值和过长值，规则与 worker 协议保持一致。 */
export function parseJudgerId(rawJudgerId: string): ParseResult<JudgerId> {
  const normalized = rawJudgerId.trim()

  if (!normalized) {
    return { ok: false, error: 'Judger id is required.' }
  }

  if (normalized.length > 120) {
    return { ok: false, error: 'Judger id must be at most 120 characters.' }
  }

  return { ok: true, value: createJudgerId(normalized) }
}
